### PR TITLE
Pull in external module dependencies when writing modulefiles

### PR DESCRIPTION
**Pull request by koysean**
Fix for https://github.com/spack/spack/issues/35212

This PR modifies spack's module APIs to partially support multiple modulefiles for a single spec. These changes allow Spack to handle an external package with multiple modulefiles.

The primary API changes are in `spack.modules.common.get_module()` and `spack.modules.common.BaseFileLayout.use_name`. The return types were changed from `str` to `List[str]` to accomodate a single spec having multiple module names.

In addition, modulefiles are no longer created for external packages with modules, and modulefiles which depend on external modules will now directly load the external modules (if autoload is enabled).

One possible issue is that the module index will not be updated with external modules. This doesn't seem like a problem to me since Spack isn't managing the external modules anyways, but I'm not very familiar with how Spack uses the module index. Is this an issue?

### CODE REVIEW

File: lib/spack/spack/cmd/modules/__init__.py [hunk 1]
```diff
@@ -169,15 +169,16 @@ def loads(module_type, specs, args, out=None):
...
-                module_type,
-                spec,
-                get_full_path=False,
-                module_set_name=args.module_set_name,
-                required=False,
-            ),
+            module,
         )
         for spec in specs
+        for module in spack.modules.common.get_module(
```
Comment (becker33) [inline lib/spack/spack/cmd/modules/__init__.py]: If this is going to return multiple modules, it should be renamed to `get_modules`.

File: lib/spack/spack/cmd/modules/__init__.py [hunk 2]
```diff
@@ -169,15 +169,16 @@ def loads(module_type, specs, args, out=None):
...
+            module_type,
+            spec,
+            get_full_path=False,
+            module_set_name=args.module_set_name,
+            required=False,
+        )
```
Comment (becker33) [inline lib/spack/spack/cmd/modules/__init__.py:179]: Oh goodness this is hard to read... if the style checker is going to format it like this we might need to have a comment explaining what's going on here.

File: lib/spack/spack/modules/common.py [hunk 3]
```diff
@@ -605,10 +614,13 @@ def dirname(self):
 
     @property
     def use_name(self):
-        """Returns the 'use' name of the module i.e. the name you have to type
-        to console to use it. This implementation fits the needs of most
-        non-hierarchical layouts.
+        """Returns a list of the 'use' names of the module i.e. the names you
```
Comment (becker33) [inline lib/spack/spack/modules/common.py:621]: Another method that should be pluralized with the changes you're making.

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/cmd/modules/__init__.py (hunk 2 above)
Comment (koysean) [reply lib/spack/spack/cmd/modules/__init__.py:179]: Added a comment - the style checker also changed its mind (?) and wants the tuple (spec, module) on one line now, so it's a lot easier to read.

**Pull request closed without merging**